### PR TITLE
 x/il: impermanent loss module

### PR DESCRIPTION
**Pull request by fedekunze**
- IL module template
-  x/il: impermanent loss module

### CODE REVIEW

File: scripts/protocgen.sh [hunk 1]
```diff
@@ -43,7 +43,7 @@ buf protoc \
 --doc_out=./docs/core \
 --doc_opt=./docs/protodoc-markdown.tmpl,proto-docs.md \
 $(find "$(pwd)/proto" -maxdepth 5 -name '*.proto')
-go mod tidy
+# go mod tidy
```
Comment (fedekunze) [inline scripts/protocgen.sh:46]: reminder to undo this change

File: x/il/keeper/abci.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+)
+
+// EndBlocker is called at the end of every block
+func (k Keeper) EndBlocker(ctx sdk.Context) {
+	// iterate over each stoploss position
+	//   fetch pair info from oracle
+	//   calculate current_ratio/reference_pair_ratio
+	//   if slippage > max_slipage
+	//       send eth transaction to withdraw lp_shares liquidity for pair_id
+}
```
Comment (fedekunze) [inline x/il/keeper/abci.go:14]: @jackzampolin @zmanian

File: proto/il/v1/genesis.proto [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+
+// GenesisState all impermanent loss state that must be provided at genesis.
+message GenesisState {
+    Params params = 1 [(gogoproto.nullable) = false];
+    repeated StoplossPosition stoploss_positions = 2 [(gogoproto.nullable) = false];
+}
+
+// StoplossPosition represents an impermanent loss stop position for a given address and uniswap pair.
+message StoplossPosition {
```
Comment (jackzampolin) [inline proto/il/v1/genesis.proto:17]: This looks good. Another way to do it would be to store them as `map<address>[]StopLoss`?

File: proto/il/v1/il.proto [hunk 4]
```diff
@@ -0,0 +1,32 @@
+syntax = "proto3";
+package il.v1;
+
+import "gogoproto/gogo.proto";
+
+option go_package = "github.com/peggyjv/sommelier/x/il/types";
+
+// Stoploss defines a set of parameters that together trigger a stoploss withdrawal.
+message Stoploss {
```
Comment (jackzampolin) [inline proto/il/v1/il.proto:9]: ++

File: proto/il/v1/query.proto [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import "gogoproto/gogo.proto";
+import "il/v1/il.proto";
+import "google/api/annotations.proto";
+import "cosmos/base/query/v1beta1/pagination.proto";
+
+
+option go_package = "github.com/peggyjv/sommelier/x/il/types";
+
+// Query defines a gRPC query service for the impermanent loss module.
+service Query {
```
Comment (jackzampolin) [inline proto/il/v1/query.proto:13]: This is great for now.

File: scripts/protocgen.sh (hunk 1 above)
Comment (jackzampolin) [reply scripts/protocgen.sh:46]: I personally like the fact that this runs w/o building. It's nice for refactoring and when being bandwidth sensitive. Maybe add this back in a make command?

File: x/il/keeper/abci.go (hunk 2 above)
Comment (jackzampolin) [reply x/il/keeper/abci.go:14]: Lets define the oracle expected keeper:

```
type OracleKeeper interface {
  GetUniswapPair(id string) Pair
}
```

[`Pair`](https://github.com/PeggyJV/sommelier/pull/13/files#diff-269b5141f20cac4c1c5e366d7a245a2b8d1832c7868ebbdc1bfa5428e1eee751R28) definition

File: x/il/module.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
+package oracle
```
Comment (jackzampolin) [inline x/il/module.go:1]: nit, might want to call this package something thats not oracle.

File: x/il/types/expected_keeper.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+// OracleKeeper is expected keeper for the oracle module
+type OracleKeeper interface {
```
Comment (jackzampolin) [inline x/il/types/expected_keeper.go:9]: `StakingKeeper`?

File: x/il/types/expected_keeper.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	Validator(ctx sdk.Context, address sdk.ValAddress) stakingtypes.ValidatorI // get validator by operator address; nil when validator not found
+	TotalBondedTokens(sdk.Context) sdk.Int                                     // total bonded tokens within the validator set
+	Slash(sdk.Context, sdk.ConsAddress, int64, int64, sdk.Dec)                 // slash the validator and delegators of the validator, specifying offence height, offence power, and slash fraction
+	Jail(sdk.Context, sdk.ConsAddress)                                         // jail a validator
+	IterateValidators(sdk.Context, func(index int64, validator stakingtypes.ValidatorI) (stop bool))
+}
+
+// EthBridgeKeeper is expected keeper for the peggy bridge module
+type EthBridgeKeeper interface {
+	AllocateTokensToValidator(ctx sdk.Context, val stakingtypes.ValidatorI, tokens sdk.DecCoins)
```
Comment (jackzampolin) [inline x/il/types/expected_keeper.go:19]: ?

File: x/il/types/keys.go [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+// Items are stored with the following key: values
+//
+// - 0x01<address_Bytes><uniswap_pair_id>: Stoploss
+var (
+	// Keys for store prefixes
+	StoplossKeyPrefix = []byte{0x01} // prefix for each key to a stoploss
+)
+
+// StoplossKey defines the full unprefixed store key for Stoploss
+func StoplossKey(address sdk.AccAddress, uniswapPair string) []byte {
```
Comment (jackzampolin) [inline x/il/types/keys.go:29]: Lets assume each address can have multiple stop loss positions. Lets make it store a `[]Pair` here.

File: proto/il/v1/tx.proto [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+option go_package = "github.com/peggyjv/sommelier/x/il/types";
+
+// MsgService defines the msgs that the oracle module handles.
+service Msg {
+  rpc CreateStoploss          (MsgStoploss)          returns (MsgStoplossResponse);
+}
+
+// MsgStoploss defines a stoploss position
+message MsgStoploss {
```
Comment (jackzampolin) [inline proto/il/v1/tx.proto:14]: Lets also define:

```proto
message MsgRemoveStopLoss {
  string address = 1;
  StopLossID = 2; 
}
```

File: x/il/types/params.go [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+)
+
+// DefaultParamspace defines default space for oracle params
+const DefaultParamspace = ModuleName
+
+// Parameter keys
+var (
+	ParamStoreKeyContractAddress = []byte("ContractAddress")
```
Comment (jackzampolin) [inline x/il/types/params.go:15]: cc @zmanian do we want to store this here? This would be the pattern moving forward for new modules (i.e. module defines it's strategy contract)

File: x/il/types/expected_keeper.go (hunk 7 above)
Comment (fedekunze) [reply x/il/types/expected_keeper.go:9]: Ah yeah need to change this

File: proto/il/v1/genesis.proto (hunk 3 above)
Comment (fedekunze) [reply proto/il/v1/genesis.proto]: we would need to iterate over the map on init genesis, so it could panic due to non-determinism.

File: x/il/types/keys.go (hunk 9 above)
Comment (fedekunze) [reply x/il/types/keys.go:29]: yup this is just the key, see the comment on the prefix above

File: app/app.go [hunk 12]
```diff
@@ -133,12 +143,12 @@ var (
...
 		govtypes.ModuleName:            {authtypes.Burner},
 		ibctransfertypes.ModuleName:    {authtypes.Minter, authtypes.Burner},
-		oracletypes.ModuleName:         nil,
+		oracletypes.ModuleName:         nil, // TODO:??
 	}
 
 	// module accounts that are allowed to receive tokens
 	allowedReceivingModAcc = map[string]bool{
-		oracletypes.ModuleName: true,
+		oracletypes.ModuleName: true, // TODO: why?
```
Comment (jackzampolin) [inline app/app.go:151]: For paying out the people who submit data w/in the reward band.

Comment (fedekunze) [reply app/app.go:151]: but this allows them to send a transfer to the module address

File: app/app.go [hunk 13]
```diff
@@ -121,7 +129,9 @@ var (
 		evidence.AppModuleBasic{},
 		transfer.AppModuleBasic{},
 		vesting.AppModuleBasic{},
+		ethbridge.AppModuleBasic{},
 		oracle.AppModuleBasic{},
+		il.AppModuleBasic{},
```
Comment (jackzampolin) [inline app/app.go:134]: :eyes: :eyes: :eyes:

File: app/app.go [hunk 14]
```diff
@@ -133,12 +143,12 @@ var (
 		stakingtypes.NotBondedPoolName: {authtypes.Burner, authtypes.Staking},
 		govtypes.ModuleName:            {authtypes.Burner},
 		ibctransfertypes.ModuleName:    {authtypes.Minter, authtypes.Burner},
-		oracletypes.ModuleName:         nil,
+		oracletypes.ModuleName:         nil, // TODO:??
```
Comment (jackzampolin) [inline app/app.go:146]: will need perms for rewarding voters, but we need to design the incentive model first.

File: app/app.go (hunk 12 above)
Comment (jackzampolin) [inline app/app.go:151]: So, re-reviewing this, could be good to allow transfers to the module account address in the case where we have a set amount of funds for paying out the oracle feeders. But this is pending the TBD economic model here.

Comment (jackzampolin) [inline app/app.go:151]: So, re-reviewing this, could be good to allow transfers to the module account address in the case where we have a set amount of funds for paying out the oracle feeders. But this is pending the TBD economic model here.

File: proto/il/v1/tx.proto [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+// MsgService defines the msgs that the oracle module handles.
+service Msg {
+  rpc CreateStoploss  (MsgStoploss)   returns (MsgStoplossResponse);
+  rpc DeleteStoploss  (MsgDeleteStoploss)   returns (MsgDeleteStoplossResponse);
+}
+
+// MsgStoploss defines a stoploss position
+message MsgStoploss {
+    string address = 1;
+    Stoploss stoploss = 2;
```
Comment (jackzampolin) [inline proto/il/v1/tx.proto:17]: One thought, maybe we want the user to not submit the reference pair ratio, and/or we check it/set it using the oracle data

Comment (jackzampolin) [inline proto/il/v1/tx.proto:17]: One thought, maybe we want the user to not submit the reference pair ratio, and/or we check it/set it using the oracle data

File: x/il/client/cli/query.go [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+	}...)
+
+	return ilQueryCmd
+
+}
+
+// GetCmdQueryStoploss implements the query rate command.
+func GetCmdQueryStoploss() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "stoploss [address] [uniswap_pair]",
```
Comment (jackzampolin) [inline x/il/client/cli/query.go:36]:
```suggestion
		Use:   "stoploss [address] [[uniswap_pair]]",
```

Comment (jackzampolin) [inline x/il/client/cli/query.go:36]:
```suggestion
		Use:   "stoploss [address] [[uniswap_pair]]",
```

File: proto/il/v1/tx.proto (hunk 15 above)
Comment (jackzampolin) [inline proto/il/v1/tx.proto:17]: One thought, maybe we want the user to not submit the reference pair ratio, and/or we check it/set it using the oracle data

File: x/il/client/cli/query.go [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+		Args:  cobra.RangeArgs(1, 2),
+		Short: "Query the stoploss positions for a given address", // TODO: update "Luna"
```
Comment (jackzampolin) [inline x/il/client/cli/query.go:38]:
```suggestion
		Short: "Query the stoploss positions for a given address",
```

Comment (jackzampolin) [inline x/il/client/cli/query.go:38]:
```suggestion
		Short: "Query the stoploss positions for a given address",
```

File: x/il/keeper/abci.go [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+)
+
+// EndBlocker is called at the end of every block
+func (k Keeper) EndBlocker(ctx sdk.Context) {
+	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), telemetry.MetricKeyEndBlocker)
+
+	params := k.GetParams(ctx)
+
+	k.IterateStoplossPositions(ctx, func(address sdk.AccAddress, stoploss types.Stoploss) (stop bool) {
+		//   TODO: fetch pair info from oracle given the pair address
```
Comment (jackzampolin) [inline x/il/keeper/abci.go:22]: Should I add a helper function for this to the oracle, or are you OK pulling the whole blob up at the beginning of end blocker?

File: x/il/client/cli/query.go (hunk 17 above)
Comment (jackzampolin) [inline x/il/client/cli/query.go:38]:
```suggestion
		Short: "Query the stoploss positions for a given address",
```

File: x/il/keeper/abci.go (hunk 18 above)
Comment (jackzampolin) [inline x/il/keeper/abci.go:22]: Should I add a helper function for this to the oracle, or are you OK pulling the whole blob up at the beginning of end blocker?

File: x/il/keeper/abci.go [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+		// log and emit metrics
+		k.Logger(ctx).Info("stoploss executed", "pair", stoploss.UniswapPairId, "address", address.String())
+
+		defer func() {
+			// amount metric
+			telemetry.SetGaugeWithLabels(
+				[]string{"stoploss", "execution"},
+				float32(stoploss.LiquidityPoolShares),
+				[]metrics.Label{telemetry.NewLabel("pair", stoploss.UniswapPairId)},
+			)
```
Comment (jackzampolin) [inline x/il/keeper/abci.go:66]: This is reminding me to add telemetry to the oracle.

File: x/il/keeper/msg_server.go [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+// MsgServer is the server API for Msg service.
+
+// CreateStoploss for a given uniswap pair
+func (k Keeper) CreateStoploss(c context.Context, msg *types.MsgStoploss) (*types.MsgStoplossResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+
+	// NOTE: error checked during msg validation
+	address, _ := sdk.AccAddressFromBech32(msg.Address)
+
+	// TODO: check if uniswap pair exists on the oracle
```
Comment (jackzampolin) [inline x/il/keeper/msg_server.go:23]:
```suggestion
	// TODO: check if uniswap pair exists on the oracle
	// TODO: ensure that pair ratio is w/in band of pair ratio from oracle
```

File: x/il/types/expected_keeper.go [hunk 21]
```diff
@@ -0,0 +1,21 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	// oracletypes "github.com/peggyjv/sommelier/x/oracle/types"
+)
+
+// OracleKeeper is expected keeper for the oracle module
+type OracleKeeper interface {
+	// GetUniswapPair(ctx sdk.Context, id string) oracletypes.Pair
```
Comment (jackzampolin) [inline x/il/types/expected_keeper.go:10]: I can add this to the oracle.

**Review by jackzampolin — APPROVED** [review]
Looks great, this also telling me some things to add/fix in the oracle.

**Comment by zmanian** [comment]
Thinking about Pair ID and the addresses of the Pair contracts and the denoms for LP tokens and where to stick some validation.

for the Pair in the create stop loss tx.

https://uniswap.org/docs/v2/smart-contract-integration/getting-pair-addresses/

File: x/il/client/cli/query.go (hunk 16 above)
Comment (fedekunze) [reply x/il/client/cli/query.go:36]: ?

File: x/il/keeper/abci.go [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+
+		//   calculate current_ratio/reference_pair_ratio
+		currentSlippage := currentRatio.Quo(stoploss.ReferencePairRatio)
+
+		if currentSlippage.LTE(stoploss.MaxSlippage) {
+			// threshold not met, continue with the next position
+			return false
+		}
+
+		// Since the current slipage is grater than max slippage we now withdraw the liquidity of the LP for the uniswap pair
```
Comment (kkennis) [inline x/il/keeper/abci.go:84]: I remember a few weeks back, we talked about different ways to model the stop-loss trigger:

* USD value - e.g., 'I put 10k into this pool, stop me out if the total position value falls below 9k'
* Token ratio - e.g. 'I put 100 UNI and 2000 DAI into this pool, stop me out if I fall below 90 UNI')

This looks like a calculation on the second bullet point - is the first bullet point covered by the `currentSlippage` calculation, or is that no longer a model of stop loss we plan to expose? 

I think we could do that with something like `usdValueOfPosition = (lpTokensOfUser/totallpTokensInPool) * reserveUSD`, but maybe that's not an interesting way of modeling LP exposure?

File: x/il/client/cli/query.go (hunk 16 above)
Comment (jackzampolin) [reply x/il/client/cli/query.go:36]: To indicate it's optional?

**Comment by jackzampolin** [comment]
@zmanian good call on adding validation on stop loss creation in the message handler. I think that would be best with 'GetPair' keeper function. Think @fedekunze has a design for how to do this in a generalized way.

File: x/il/keeper/abci.go (hunk 22 above)
Comment (jackzampolin) [reply x/il/keeper/abci.go:84]: we might be mixing the two here @kkennis. Will read through again tomorrow.

File: x/il/keeper/abci.go [hunk 23]
```diff
@@ -51,30 +51,52 @@ func (k Keeper) EndBlocker(ctx sdk.Context) {
...
+		// 	return false
+		// }
+
+		totalSupply, err := sdk.NewDecFromStr(totalSupplyStr)
 		if err != nil {
 			return false
 		}
 
-		reserve1, err := sdk.NewDecFromStr(reserveStr1)
+		reserveUSD, err := sdk.NewDecFromStr(reserveUSDStr)
```
Comment (kkennis) [inline x/il/keeper/abci.go:108]: Sorry to go back and forth on this 😀 but is there a reason we need to support one or the other (one being 'price ratio', the other being 'usd value'). Can we support both?

Don't know how the rest of the protocol works but I'm assuming this would require a new field on `stoploss`, something like `stoploss.calculationType`. That could have values of `priceRatio` or `usdValue`. Then depending on the value of `calculationType` you would calculate slippage based on price ratio or based on usd value.

My follow-up question would be - let's say your stoploss is at 10%. Does a 10% change in the price ratio of the two pairs equate to a 10% loss of usd value of the position? Definitely not, especially if neither token in the pair is a stablecoin. 

Some back of the napkin math...
```
1 WETH to 1000 USDC - 2000 USD liquidity
…WETH price goes up to 1500 USDC….
Price ratio is 1500 WETH to 1 USDC - 50% price ratio change
This is a 2% IL loss (realized loss is smaller according to fees)
```

And we can see from https://uniswap.org/docs/v2/advanced-topics/understanding-returns/ that the relationship between price ratio and IL is plotted by this graph:

Screen Shot 2021-02-08 at 11 06 45 AM

So, question is - can we give the user a way to model their loss in _either_ way - price ratio OR usd value - and then use the relationship in the screenshot above to figure out their loss in their preferred model?

Don't want to overcomplicate things, if we have to pick one approach, I think USD value is best, so this code looks good. But wondering if this type of flexibility is possible b/c that's how I was thinking about the UI interaction, the user would get a chance to say either:

* 'stop me out when I lose X USD' OR
* 'stop me out when I get rebalanced Y far into one coin' (this is equivalent to price ratio)

And the modeled relationship above is what I think we need to achieve that

Comment (kkennis) [reply x/il/keeper/abci.go:108]: Basically this is the relationship:

```
initialPriceRatio = stoploss.ReferencePairRatio
currentPriceRatio = pair.reserve0 / pair.reserve1
priceRatioChange = initialPriceRatio / currentPriceRatio
impermanentLossRatio = 2 * sqrt(priceRatioChange) / (1+priceRatioChange) - 1 
impermanentLoss = impermanentLossRatio * usdValueOfPosition
totalReturn = stoploss.accruedFees - impermanentLoss (not sure if we have accrued fees)
```

Then, depending on how the user chose to model it, we either see if totalReturn is less than 90% of their initial usd value of their position, or if priceRatioChange is over 10%. (assuming 10% as the 'stoploss baseline').

Comment (fedekunze) [reply x/il/keeper/abci.go:108]: Thanks @kkennis for the detailed comment

> -  'stop me out when I lose X USD' OR
- 'stop me out when I get rebalanced Y far into one coin' (this is equivalent to price ratio)

I think the first one is what users would care about the most in the beginning. But I'm def open to integrate both options, but I'd prefer to do the second one once wire the IL module with the smart contract and oracle logic and refactor the oracle module. Would that be ok? I'm open to suggestions

File: x/il/keeper/abci.go [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+
+		if currentSlippage.LTE(stoploss.MaxSlippage) {
+			// threshold not met, continue with the next position
+			return false
+		}
+
+		// Since the current slipage is grater than max slippage we now withdraw the liquidity of the LP for the uniswap pair
+		// that is suffering impermanent loss
+
+		uniswapERC20Pair := bridgetypes.NewERC20Token(uint64(stoploss.LiquidityPoolShares), stoploss.UniswapPairId)
```
Comment (jtremback) [inline x/il/keeper/abci.go:91]: Why is there a uint64 in here?

File: x/il/keeper/abci.go [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+	params := k.GetParams(ctx)
+
+	invalidationID := k.GetInvalidationID(ctx)
+
+	// variable for the original value to check if we need to store the new invalidationID value
+	originalInvalidationID := invalidationID
+
+	pairMap := make(map[string]*oracletypes.UniswapPair)
+
+	k.IterateStoplossPositions(ctx, func(address sdk.AccAddress, stoploss types.Stoploss) (stop bool) {
```
Comment (marbar3778) [inline x/il/keeper/abci.go:45]: It would be interesting to do some benchmarking on this later on. Depending on the usage we could run into similar issue as kava with validators missing blocks because they have too much in the endblock/beginblock.

Comment (fedekunze) [reply x/il/keeper/abci.go:45]: Oh wait until you see the Peggy module endblocker, that one is large

Comment (zmanian) [reply x/il/keeper/abci.go:45]: Sommelier is expected to be a very resource intensive process for validators to run.
